### PR TITLE
system: Add support to configure timedated via timedatectl

### PR DESCRIPTION
**Pull request by ssahani**
##### SUMMARY

This work automates timedatectl

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

timedatectl
##### ANSIBLE VERSION

```paste below
"ansible 2.8.0.dev0"
```

##### ADDITIONAL INFORMATION

```paste below
          - timedatectl:                                                                                                                                                                         
             timezone: "Asia/Kolkata"                                                                                                                                                      
             ntp: yes 
```

Signed-off-by: Susant Sahani <susant@redhat.com>

### CODE REVIEW

File: lib/ansible/modules/system/timedatectl.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+module: timedatectl
+short_description: Automates timedatectl
+description:
+    - Allows you to configure timedated.
+version_added: "2.8"
+options:
+    command:
+        description:
+            - Specifies the command that should be applied.
+        choices: [ "set-time", "set-timezone", "set-local-rtc", "set-ntp" ]
```
Comment (bcoca) [inline lib/ansible/modules/system/timedatectl.py]: we normally prefer establishing a 'state' not passing commands

Comment (ssahani) [reply lib/ansible/modules/system/timedatectl.py]: Thanks updated.

File: lib/ansible/modules/system/timedatectl.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            time=dict(required=False, type=str),
+            timezone=dict(required=False, type=str),
+            ntp=dict(required=False, type=bool),
+            local_rtc=dict(required=False, type=bool),
+        ),
+        supports_check_mode=True
```
Comment (alxgu) [inline lib/ansible/modules/system/timedatectl.py]: should be false. It doesn't support it now

Comment (ssahani) [reply lib/ansible/modules/system/timedatectl.py]: yes updated.

**Comment by tmshn** [comment]
Some thoughts.

- `timezone`, `local_rtc` → same thing can be achieved by `timezone` module. I thinki it's not a good idea to confuse users.
- `time` → IMO ansible playbooks should be idempotent, and setting time via Ansible doesn't seem rational.

So how about adding simple `ntp` module like this:

```yaml
ntp:
  enabled: true
```

Maybe it's ok to only support systems where `timedatectl` command is available (as a first step).

**Comment by tmshn** [comment]
Or even `ntp: true` might be able to be replaced by existing module...

```yaml
service:
  name: systemd-timesyncd # or chrony or ntpd
  state: started
  enabled: true
```

**Comment by ssahani** [comment]
I don't think there is any confusion here. When you take a namespace and a module you already know what you trying to do. By saying that there are multiple configuration can be done via different modules . I failed to figure where is you confusion and and why time can't be set ?

**Comment by tmshn** [comment]
Sorry, my explanation might be bit poor. What I meant was: there are already modules to achieve same things as this module does (except for `time` parameter). Therefore I don't understand the necessity or pros of this module.

This

```yaml
- timedatectl:
    timezone: Asia/Kolkata
    ntp: yes
    local_rtc: yes
```

is equivalent to this

```yaml
- timezone:
    name: Asia/Kolkata
    rtc: local
- service:
    name: systemd-timesyncd
    state: started
    enabled: true
```

_N.B.: I'm not the maintainer, so it's not me who decides if this gets merged. This is just my humble opinion._

**Comment by tmshn** [comment]
And for `time` parameter, I can't imagine the usage.

Do you write down a time literally like this?

```yaml
- timedatectl:
    time: 2018-11-23 15:13:21
```

Then, your target machine's clock is set to the time **every time you run the playbook**.

Or you pass value via variable from somewhere else (like previous step etc.)??

```yaml
- timedatectl:
    time: "{{ time_from_xxx }}"
```

This could be okay, but be aware that there could be a few seconds of lag due to ansible's communication through ssh connection.

**Comment by ssahani** [comment]
> Sorry, my explanation might be bit poor. What I meant was: there are already modules to achieve same things as this module does (except for `time` parameter). Therefore I don't understand the necessity or pros of this module.
> 
> This
> 
> ```yaml
> - timedatectl:
>     timezone: Asia/Kolkata
>     ntp: yes
>     local_rtc: yes
> ```
> 
> is equivalent to this
> 
> ```yaml
> - timezone:
>     name: Asia/Kolkata
>     rtc: local
> - service:
>     name: systemd-timesyncd
>     state: started
>     enabled: true
> ```
> 
> _N.B.: I'm not the maintainer, so it's not me who decides if this gets merged. This is just my humble opinion._

I already replied to this .

**Comment by ssahani** [comment]
> And for `time` parameter, I can't imagine the usage.
> 
> Do you write down a time literally like this?
> 
> ```yaml
> - timedatectl:
>     time: 2018-11-23 15:13:21
> ```
> 
> Then, your target machine's clock is set to the time **every time you run the playbook**.
> 
> Or you pass value via variable from somewhere else (like previous step etc.)??
> 
> ```yaml
> - timedatectl:
>     time: "{{ time_from_xxx }}"
> ```
> 
> This could be okay, but be aware that there could be a few seconds of lag due to ansible's communication through ssh connection.

See https://www.freedesktop.org/software/systemd/man/timedatectl.html

**Comment by tmshn** [comment]
> I already replied to this .

Sorry, my English reading skill is too low to understand your reply.
You meant "it's ok to add a new module which is similar to existing ones and Ansible users won't confuse which one to use"?

> See https://www.freedesktop.org/software/systemd/man/timedatectl.html

I should say "use cases" rather than "usage". What I want to know is more real-life situation/scenario to set time using Ansible.

**Comment by ssahani** [comment]
> > I already replied to this .
> 
> Sorry, my English reading skill is too low to understand your reply.
> You meant "it's ok to add a new module which is similar to existing ones and Ansible users won't confuse which one to use"?

Let me get this strate why there are multiple software doing the same stuff ?

> 
> > See https://www.freedesktop.org/software/systemd/man/timedatectl.html
> 
> I should say "use cases" rather than "usage". What I want to know is more real-life situation/scenario to set time using Ansible.

May be you don;t have . I do have for sure. And the logic for ssh delay how you are doing now ?

**Comment by tmshn** [comment]
Sorry I was not able to understand your questions...anyway, I understand you have a concrete demand for this module. Then I have nothing to say any more.

**Pull request closed without merging**